utils/emmylua/gui.lua

```diff
@@ -0,0 +1,226 @@
+---@param id string An ID for the definition, to be referenced from WML
+---@param content WML The definition WML
+function gui.add_widget_definition(type, id, content) end
+
+---A reference to a widget in a custom dialog box
+---@class widget : gui.widget
+---@field enabled boolean
+---@field tooltip tstring
+---@field visible boolean|"'visible'"|"'hidden'"|"'invisible'"
+---@field type string
+---@field on_left_click fun()
+
+---The window widget is a container that contains all other widgets in the dialog
+---@class window : widget
+
+---A simple widget, with no children
+---@class simple_widget : widget
+---@field use_markup boolean
+---@field label tstring
+---@field marked_up_text tstring
+
+---A button with two or more states, typically used as a checkbox or radiobutton
+---@class toggle_button : simple_widget
+---@field selected boolean
+---@field selected_index integer
+---@field on_modified fun()
+
+---Similar to a toggle button, but allows for arbitrary contents
+---@class toggle_panel : widget
+---@field selected boolean
+---@field selected_index integer
+---@field on_modified fun()
+
+---A dynamic list of items, shown with a scrollbar
+---@class listbox : widget
+---@field selected_index integer
+---@field item_count integer
+---@field on_modified fun()
+
+---A container widget that shows only one of its children at any given time
+---@class multi_page : widget
+---@field selected_index integer
+---@field item_count integer
+
+---A container widget that whose children all occupied the same space, overlayed on top of each other
+---@class stacked_widget : widget
+---@field selected_index integer
+
+---A button that produces a dropdown menu when clicked
+---@class menu_button : widget
+---@field selected_index integer
+---@field on_modified fun()
+
+---An editable text box
+---@class text_box : widget
+---@field text string
+---@field on_modified fun()
+
+---A slider
+---@class slider : widget
+---@field value integer
+---@field on_modified fun()
+
+---A progress bar
+---@class progress_bar : widget
+---@field percentage number
+
+---A dynamic, hierarchical list of items, shown with a scrollbar
+---@class treeview : widget
+---@field selected_item_path integer[]
+---@field on_modified fun()
+
+---A single node in a tree view
+---@class tree_view_node : widget
+---@field path integer[]
+---@field unfolded boolean
+
+---A panel that shows details on a given unit or unit type
+---@class unit_preview_pane : widget
+---@field unit unit|unit_type
+
+---A simple button that triggers once when clicked
+---@class button : simple_widget
+---@field on_button_click fun()
+
+---A static text label
+---@class label : simple_widget
+---A simple image
+---@class image : simple_widget
+---A simple button that triggers repeatedly if the mouse is held down
+---@class repeating_button : button
+---A generic container that lays out its children in a grid
+---@class grid : widget
+
+---@class gui.widget
```

Comment:
I'm not confident that the above covers every possible type of widget… can anyone see one I've missed?